### PR TITLE
4.0 beta

### LINKED ISSUE #210: Download an image and send it to chat when it's url?
I have an idea - if there's function call like this

``` javascript
api.sendImage("url", "http://example.com/image.jpg", "", event.thread_id);
```

to the facebook integration API (and maybe some others?), it could download the image and send it to the thread as an attachment instead of just sending the link. If someone wants to send a link instead of image, he can always use `sendMessage` function. I think it shouldn't be that hard to implement. There could be some checks to ensure the file to download is not too big.

### LINKED ISSUE #217: Make Integrations and Services installable modules
- Allow integrations to be installed, uninstalled and configured at runtime
- Allow modules to be 'services' (that is, no I/O, only start/stop)
- Add onload/unload module events - all ES6 code should extend EventEmitter so we get events by default.  Useful for services.
- Allow I/O interception (middleware). Also useful for services.

Reasoning:
- Feature parity and requests
- Makes Concierge [even more] modular
- Reduces complexity of Concierge core
- Events reduce dependence on implementation, only on emitted data.
- Allows modules such as [boss](http://github.com/concierge/boss) to manage the Concierge platform without modifying it.

### PR DESCRIPTION
**Pull request by mrkno**
Resolves #217, resolves #210 (via middleware)

Todo:
- ~~Allow integrations to be installed, uninstalled and configured at runtime~~
- ~~Allow modules to be 'services' (that is, no I/O, only start/stop)~~
- ~~Add onload/unload module events - all ES6 code should extend EventEmitter so we get events by default.~~
- ~~Allow I/O interception (middleware). Also useful for services.~~
- ~~Gracefully handle non-existent integrations~~
- ~~Fix help~~ https://github.com/concierge/help/pull/1
- ~~Fix admin~~ https://github.com/concierge/admin/pull/1
- ~~Fix disable~~ https://github.com/concierge/disable/pull/1
- ~~Fix lolphp~~ https://github.com/concierge/lolphp/pull/1
- ~~Fix programmer~~ https://github.com/concierge/programmer/pull/1
- ~~Fix pickup~~ https://github.com/concierge/pickup/pull/1
- ~~Fix joke~~ https://github.com/concierge/joke/pull/1
- ~~Fix insult~~ https://github.com/concierge/insult/pull/1
- ~~Fix quote~~ https://github.com/concierge/quote/commit/8a695888dea157964cfc6a2d25eb20811861a0a6
- ~~Fix update~~ https://github.com/concierge/update/pull/1
- ~~Update KPM~~ https://github.com/concierge/kpm/pull/6
- ~~Documentation updates~~
- ~~Module unloading infinite loop when a failure occurs~~
- ~~Config as an unloadable/overrideable service~~
- ~~Integration config migration~~

Wishlist:
- ~~Async module loading~~ deferred to 4.1
  - ~~Decoupled GUI~~
  - ~~Async loading calls~~
- ~~Prevent Facebook crashes from bringing down Concierge~~
- ~~Implement means to blame a module for being the source of an exception~~
- ~~Reference counting to deprecate `require.once` and `require.safe`~~

Fixed:
- Local files that could not be resolved would be installed via NPM
- Loopback would not work unless using ES6 style integrations

### CODE REVIEW

**Comment by mrkno** [comment]
@phillipadsmith @ParasDPain @Awarua- @p0358 if there are any additional changes you want made for the beta branch other than what is still TODO, now would be a good time to mention them.

**Comment by ParasDPain** [comment]
My wishlist atm is just persistent heroku and a twitter listener....and I think both can be done with the current functionality in Concierge?

**Comment by phillipadsmith** [comment]
@mrkno Thank you for the ping and patience with my response. Finally had some time to dive back into this today. The list above looks great. Nothing to add at the moment.

**Comment by p0358** [comment]
@mrkno Maybe a bit late, but nice feature would be integration break detection. Last time `facebook-chat-api` likes to break oftenly and then Facebook bot just doesn't work after the error until Concierge is restarted. It could try to automatically reload integration instead. Or `facebook` integration alone could get something like that.

**Comment by mrkno** [comment]
@phillipadsmith cool, just file an issue/create a further PR if you retrospectively think of anything.

@ParasDPain I'll see what I can do. No promises... I'm also slowly looking into the `facebook-chat-api` problem but it takes aaaaaggggeees to reproduce.

**Comment by ParasDPain** [comment]
@mrkno Just mentioning, the documentation can be stylistically updated for the [integration configurations](https://github.com/concierge/Concierge/tree/master/doc/integrations) to be consistent with the rest of the docs but if integrations will be moved into their own repos after this PR's merge, then these changes are not important. 

Tutorials will be a good addition as well but I'd need this PR to be merged before I can accurately write them, so they are better off in their own PR.

**Comment by mrkno** [comment]
@ParasDPain I was thinking of either redirecting the Wiki to the repo documentation (probably moving the KPM table to the KPM repo which makes sense) or the documentation in the repo to the Wiki. It would reduce the duplication.

**Comment by ParasDPain** [comment]
@mrkno I fully agree. A common source will be easier to maintain as well. Keeping the docs within the source code allows quick offline access to the docs than keeping them in the Wiki.

And the KPM table will be within the kpm module's wiki? or within it's source code? It'll be difficult to update within the source code.

Speaking of modules, I noticed `_FirstRun()` within `core/platform.js` has hard-coded module links. It'd be great if they are loaded from a separate config file somewhere so it's easier to customize installs during deploys as well.

**Comment by mrkno** [comment]
@ParasDPain a config file, or similar to the KPM table (some common but changeable online table which is referenced). You already require network access to be able to install the modules anyway so I don't see any harm pointing at an online table other than complexity.

**Comment by ParasDPain** [comment]
@mrkno Fair enough but if the table is a config file, all new entries will generate a lot of PRs. Although, a JSON syntax definitely makes it easier to parse and maintain.

**Comment by mrkno** [comment]
@ParasDPain there is now a means to override the default installs, as per https://github.com/concierge/Concierge/blob/a2fede44c300f4fd8b81fe104339b47fe64435f0/doc/DefaultCommands.md#custom-defaults

**Comment by ParasDPain** [comment]
@mrkno Great! This will make managing the broken heroku build much more easier.

File: core/modules/hubot/robot.js [hunk 1]
```diff
@@ -1,352 +1,404 @@
...
+        }
+        else {
+            this._descriptor.type = ['module'];
+        }
+    }
+
+    _inferCalleeData() {
+        // Welcome to hack land, where hacks are common place.
+        // We access event_source and thread_id from the stack trace so that this is thread safe.
+        const stack = global.getStackTrace();
```
Comment (Awarua-) [inline core/modules/hubot/robot.js]: I'm not even going to bother to work out how this works exactly

File: core/platform.js [hunk 2]
```diff
@@ -117,22 +144,29 @@ class Platform extends EventEmitter {
...
         console.title(figlet.textSync(this.packageInfo.name.toProperCase()));
-
-        console.title(' ' + this.packageInfo.version);
+        console.title(` ${this.packageInfo.version}`);
         console.info('------------------------------------');
         console.warn($$`StartingSystem`);
 
         // Load system config
         console.warn($$`LoadingSystemConfig`);
+        this.modulesLoader.loadModule({
```
Comment (Awarua-) [inline core/platform.js]: Assuming this will be extended to allow for other configuration services in the future

Comment (mrkno) [inline core/platform.js]: Check the --configserv option

File: core/modules/hubot/robot.js (hunk 1 above)
Comment (mrkno) [inline core/modules/hubot/robot.js]: Good idea. Its a hack

**Pull request merged**